src/FilenameParsing/FileResolutionStrategy.php

```diff
@@ -0,0 +1,61 @@
+<?php
+namespace SilverStripe\Assets\FilenameParsing;
+
+use League\Flysystem\Filesystem;
+
+/**
+ * Represents a strategy for resolving files on a Flysystem Adapter.
+ */
+interface FileResolutionStrategy
+{
+    /**
+     * Try to resolve a file ID against the provided Filesystem.
+     * @param string $fileID
+     * @param Filesystem $filesystem
+     * @return ParsedFileID|null Alternative FileID where the user should be redirected to.
+     */
+    public function resolveFileID($fileID, Filesystem $filesystem);
+
+    /**
+     * Try to resolve a file ID against the provided Filesystem looking at newer version of the file.
+     * @param string $fileID
+     * @param Filesystem $filesystem
+     * @return ParsedFileID|null Alternative FileID where the user should be redirected to.
+     */
+    public function softResolveFileID($fileID, Filesystem $filesystem);
+
+    /**
+     * Try to find a file ID for an existing file the provided file tuple.
+     * @param array|ParsedFileID $tuple
+     * @param Filesystem $filesystem
+     * @param boolean $strict Whatever we should enforce a hash check on the file we find
+     * @return ParsedFileID|null FileID
+     */
+    public function searchForTuple($tuple, Filesystem $filesystem, $strict = true);
```

Comment:
So this accepts a `ParsedFileID` of a file that might not exist in this path, and then returns a matching one that exists in a different path?

Reply:
Can you define how a "file tuple" is different from a "file id"? Wondering if this method is named a bit too confusing

Reply:
Many methods expect either a file ID string, a ParsedFileID or file tuple array . File tuple arrays can easily be converted to ParsedFileID because they contain the same info.

My original idea was that logic that was using file tuple arrays could easily call a FileResolutionStrategy without needing to convert the data to ParsedFileID. In hindsight, that wasn't nearly as helpful as I thought it would be.

Reply:
I suggest we stop using `$fileID` as an argument, and just generate it from the tuple as needed.

We don't really need both tuple and ParsedFileID. I don't think the class adds much value over the array as a structure, but picking one is more important than picking the right one. ;P

Reply:
I've been prioritising using Filename/Hash combo over file ID. There's a few places like `getResponseFor` where we have to use the FileID because that's what we get from the outside world.

There's a few existing methods on the `AssetStore` interface that return file tuple array. We can't change the return format of those without breaking semver.

I do think ParsedFileID provides value beyond a file array tuple. There's a lot of places where I need to swap value in and out (e.g.: getting a variantless version a file). If I'm using arrays for that, I've got to copy the values to a different array and unset the variant part. There's also the mundane problems that if you keep writing array keys as string all over the place, you're bound to mess the spelling a few times and those mistake might not be immediately apparent.

Having an immutable class to track the file tuple data makes these problems considerably easier to manage.

Reply:
> I do think ParsedFileID provides value beyond a file array tuple.

Yep, I agree - the data type benefits from a "schema" (via a PHP interface), and being able to make it immutable makes the code a lot cleaner (as opposed to arrays). But yeah, let's keep that discussion going on the main "thread" in this PR, resolving this convo and linking to it for reference from the main thread.